### PR TITLE
CNP-1131 Activate ACR image build

### PR DESCRIPTION
**Pull request by slyg**
# Description

This PR leverages on layers re-use to reduce the the docker image build time in CI.

Jira ticket: https://tools.hmcts.net/jira/browse/CNP-1131

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

The PR CI shows a decrease of build time of the docker image (1:40-ish min)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: kubernetes/deployment.template.yaml [hunk 1]
```diff
@@ -37,10 +37,10 @@ spec:
         # These may be adjusted, but be mindful please
         resources:
           requests:
-            memory: "64Mi"
+            memory: "256Mi"
```
Comment (slyg) [inline kubernetes/deployment.template.yaml]: Not sure this is something you encounter sometimes, but I had to increase those values to have a more stable deployment.

File: Dockerfile [hunk 2]
```diff
@@ -1,10 +1,12 @@
-FROM node:8.12.0-stretch
-
-WORKDIR /opt/app
-
-COPY . /opt/app
-RUN yarn --production && yarn setup && yarn cache clean
-
-CMD [ "yarn", "start" ]
-
+FROM hmcts.azurecr.io/hmcts/base/node/stretch-slim-lts-8 as base
```
Comment (hosainnet) [inline Dockerfile]: This is not publicly accessible (without authorisation, at least). Personally it seems a step backward from being able to clone and `docker build .` given that this project is open source. Is it done for caching reasons/customisation?

File: kubernetes/deployment.template.yaml (hunk 1 above)
Comment (hosainnet) [reply kubernetes/deployment.template.yaml]: Yeah, 64Mi was probably too optimistic 👍😃

File: Dockerfile (hunk 2 above)
Comment (slyg) [reply Dockerfile]: Hi, yes it is done for that reason, ACR's best practices advise to use [network-close](https://docs.microsoft.com/en-us/azure/container-registry/container-registry-best-practices#network-close-deployment) registries. Also, we intend to use ACR's ability to automatically apply security patches.

Now we can still deploy those base images in docker hub as well for open-source users. In any case, those images are only used in our CI which is not public afaik.

Comment (slyg) [reply Dockerfile]: Though OSS is something we haven't thought about I admit, are there some guidelines regarding open-source policy? So we can have this in mind in the future.

Comment (hosainnet) [reply Dockerfile]: I'm not sure sure re the policy tbh, but I guess from a developer experience, I could just clone and run the app without having to install node, yarn etc

Comment (slyg) [reply Dockerfile]: I pushed modifications in the readme to make the instructions clearer. The existing ones where using docker-compose but as there is no docker-compose.yaml file in this project they were invalid anyhow.

The node base images are a free available project.

File: kubernetes/deployment.template.yaml (hunk 1 above)
Comment (slyg) [reply kubernetes/deployment.template.yaml]: Seems memory doesn't cost much so these numbers can be applied quite extensively to other projects, especially nodejs ones. CPU on the other hand might be more constrained 😈.

**Comment by slyg** [comment]
Hi @mrganeshraja if someone could help us merge or reject this PR that would be great, thx. It's been opened for 2 weeks now and can't be merged by us.

**Pull request merged**